docs/design/app/states.md

```diff
@@ -0,0 +1,204 @@
+modification to the tree permitted after the fast copy is made, and that is to apply the hashes to the nodes in
+the tree during the hash phase).
+
+The next-most recent version of the tree is the "latest immutable state". This is the state used for answering queries
+made by customers of the consensus nodes. It is also the state used in `pre-handle` (more on this later) for preparing
+asynchronously, and ahead of time, transactions for processing.
+
+Another older state known as the "latest signed state" is the one currently being gossiped to all other nodes in the
+network. And finally, if the node is helping to "reconnect" another node, it will keep a copy of state in memory from
+which it "teaches" the other node about the state.
+
+While the "merkle tree" is a complex data structure, the service implementations actually don't need to know
+or care about it at all. The merkle tree is an implementation detail of the application, not of the services. If each
+service had to know about all these different states and which state it should use for which code path, the code would
+become quite complex and prone to mistakes! Instead, we let the Hedera application be responsible for understanding this
+complexity. Each service needs only a simple key-value store for storing data. So we created the `ReadableState` and
+`WritableState` classes as simple key-value store APIs which services can work with, and let the Hedera application 
+implement these interfaces over the right version of the merkle tree for the right calls. From the service's
+perspective, it doesn't know anything about any of these `Round`s or `Event`s or `SwirldTransaction`s or fast copies or
+mutable and immutable states. It is simply given, by the application, the `ReadableState` to use for reading or the
+`WritableState` for writing.
+
+## States
+
+There are two "state" interfaces: `ReadableState` and `WritableState`. `WritableState` extends from `ReadableState`,
+so a `WritableState` is read/write. Each service implementation has one or more `MerkleMap`s or `VirtualMap`s holding
+the data for that service. These are wrapped by the platform and made available to the service as a `WritableState` or
+`ReadableState`.
+
+Each `State` is scoped to a particular service implementation. They are never shared across service implementations.
+In fact, the `State` of a service **MUST NOT** be exposed as public API by any service API. The state of a service
+is strictly an implementation detail of the service.
+
+From the `ReadableState` a service implementation can:
+ - `get` values from the k/v store
+ - check whether a given key is contained by the state
+
+From the `WritableState` a service implementation can perform all tasks on a `ReadableState`, plus:
+ - `getForModify` to read the value for the purpose of subsequently modifying it (there is an optimized code
+   path for this case)
+ - `put` a new value
+ - `remove` a value
+ - get keys of all modified values with `modifiedKeys`
```

Comment:
Is this for use in constructing the final record?

Reply:
Yes, exactly. The idea was that there would be a special method on the crypto transaction handler that was the "post processing" handler, and it would need to be able to get from the AccountStore every account that was touched during transaction handling, so it could record all the balance changes.